_posts/2024/2024-02-11-dropping-codegen.md

```diff
@@ -0,0 +1,95 @@
+---
+title: "Dropping Project Support for Code Generation"
+date: 2024-02-11 09:00:00 +1200
+tags: [json-schema, codegen]
+toc: true
+pin: false
+---
+
+Some time ago, I released [my first attempt at code generation](/posts/exploring-codegen) from JSON Schemas.  However, I've decided to deprecate the library in favor of _Corvus.JsonSchema_.
+
+When I created _JsonSchema.Net.CodeGeneration_, I knew about _Corvus.JsonSchema_, but I thought it was merely an alternative validator.  I didn't truly understand its approach to supporting JSON Schema in .Net.
+
+Today we're going to take a look at this seeming competitor to see why it actually isn't one.
+
+## What is _Corvus.JsonSchema_?
+
+_Corvus.JsonSchema_ is a JSON Schema code generator that bakes validation logic directly into the model.
+
+To show this, consider the following schema.
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "foo": {
+      "type": "integer",
+      "minimum": 0
+    }
+  }
+}
+```
+
+As one would expect, the library would generate a class with a single property: `int foo`.  But more than mere auto-properties, it generates extra code in the setter to ensure that the model stays within the constraints expressed in the schema, even at runtime.
+
+This means that the setter for `foo` would also contain logic similar to
+
+```c#
+if (value < 0)
+    throw new ArgumentException(nameof(value), "Value must be greater than 0");
```

Comment:
This is not actually true - you can set the value but then the result won't be `IsValid() == true` any more. We avoid exceptions wherever possible because they are *so* expensive.

But the setters are strongly typed (including taking into account e.g. `format`.

Reply:
Okay.  I didn't realize that.

Looking at some [docs](https://github.com/corvus-dotnet/Corvus.JsonSchema/blob/main/docs/GettingStartedWithJsonSchemaCodeGeneration.md#validation), it seems that it will take any (typed) value, and then validation is performed in the `.IsValid()` method, as you mentioned.

Alright.  I'll update that.  (I thought the models was maintained as valid.)